collector/pipelinerun.go

```diff
@@ -2,27 +2,57 @@ package collector
 
 import (
 	"context"
-	"github.com/go-kit/log/level"
+	"fmt"
+
 	"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1"
-	"os"
 
-	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/record"
 	ctrl "sigs.k8s.io/controller-runtime"
 	"sigs.k8s.io/controller-runtime/pkg/client"
-	"sigs.k8s.io/controller-runtime/pkg/metrics"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
 )
 
-func SetupPipelineRunCachingClient(mgr ctrl.Manager) {
-	// we do not set up a reconciler to get a caching client, but we make one list
-	// call up top to get the ball rolling; confirmed via break points in a debugger that
-	// k8s methods for creating a shared informer and watch are called
-	prList := v1beta1.PipelineRunList{}
-	opts := &client.ListOptions{Namespace: metav1.NamespaceAll}
-	mgr.GetClient().List(context.Background(), &prList, opts)
-	psCollector, err := NewCollector(logger, mgr.GetClient())
+type ReconcilePipelineRun struct {
+	client        client.Client
+	scheme        *runtime.Scheme
+	eventRecorder record.EventRecorder
+	psCollector   *PipelineServiceCollector
+}
+
+func SetupPipelineRunCachingClient(mgr ctrl.Manager) error {
+	reconciler := &ReconcilePipelineRun{
+		client:        mgr.GetClient(),
+		scheme:        mgr.GetScheme(),
+		eventRecorder: mgr.GetEventRecorderFor("MetricExporterPipelineRuns"),
+		psCollector:   NewCollector(),
+	}
+	return ctrl.NewControllerManagedBy(mgr).For(&v1beta1.PipelineRun{}).Complete(reconciler)
```

Comment:
Thought - to avoid duplication, I think this needs a call to [.WithEventFilter](https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/builder#Builder.WithEventFilter) so we only capture the right state transition. That way we don't need the check for `isDone` vs `isRunning` - we just need to see if we transitioned from "Created" to "Running"

Do you agree here?

Reply:
ooh, if I read the godoc correctly, yes, that looks better... i.e. the godoc says `For example, filtering on whether the resource version has changed.`
 
I'll try it out, see if I can filter on the start time changing

Reply:
and yeah, I now remember using such filters when directly writing to shared informers and the k8s client API ... makes sense controller runtime would expose

thx for reminder / tip @adambkaplan